### PR TITLE
Separate reading of files from posting data, parallelize posting data

### PR DESCRIPTION
**Pull request by VibhanshuBhardwaj**

### CODE REVIEW

File: csv-extractor/csv_extractor.py [hunk 1]
```diff
@@ -78,7 +80,7 @@ def create_data_points(values, timestamps):
     data_points = []
 
     for i, value_string in enumerate(values):
-        if pandas.notnull(value_string):
+        if value_string != '':
```
Comment (eventh) [inline csv-extractor/csv_extractor.py]: pandas.notnull has more checks that might be important here.
I still think we should get rid of it, but we must replicate the ones that are relevant. Maybe check also for None, "null", "NULL"?

File: csv-extractor/csv_extractor.py [hunk 2]
```diff
@@ -100,29 +102,46 @@ def create_time_series(client, name: str, external_id: str) -> None:
...
 
 
+def get_parsed_file(path) -> Dict[str, str]:
+    """Parse the csv file and return the data in dictionary"""
+    parsed_file = {}
+    with open(path, "r", encoding="latin-1") as f:
+        data = csv.DictReader(f, delimiter=";")
+        for row in data:
+            for (k, v) in row.items():
+                try:
```
Comment (eventh) [inline csv-extractor/csv_extractor.py]: use defaultdict instead of try/except

File: csv-extractor/csv_extractor.py [hunk 3]
```diff
@@ -100,29 +102,46 @@ def create_time_series(client, name: str, external_id: str) -> None:
     _log_error(client.time_series.post_time_series, [new_time_series])
 
 
+def get_parsed_file(path) -> Dict[str, str]:
+    """Parse the csv file and return the data in dictionary"""
+    parsed_file = {}
+    with open(path, "r", encoding="latin-1") as f:
+        data = csv.DictReader(f, delimiter=";")
+        for row in data:
+            for (k, v) in row.items():
```
Comment (eventh) [inline csv-extractor/csv_extractor.py]: No need for parentheses around k, v :)

File: csv-extractor/csv_extractor.py [hunk 4]
```diff
@@ -131,20 +150,29 @@ def process_csv_file(client, monitor, csv_path, existing_time_series):
...
     monitor.successfully_processed_files_gauge.set(0)
-
+    network_threads = []
     for path in paths:
         try:
             try:
-                data_points_count, external_ids_count = process_csv_file(client, monitor, path, time_series_cache)
+                count_data_points, count_ext_ids, threads = process_csv_file(client, monitor, path, time_series_cache)
+                network_threads += threads
+                # post the requests if more than 20 or if we're on the last file
```
Comment (eventh) [inline csv-extractor/csv_extractor.py]: The current logic is that we delete the file after everything was posted successfully. That way we guarantee we don't lose data if there was an error. If error we move the file if that option is enabled.

This new logic will delete files before we send data, and therefor will have no way of preserving it. Sending data after each request has the drawback that we can't parallelize as well as we wish. What do you think?

File: csv-extractor/csv_extractor.py (hunk 1 above)
Comment (VibhanshuBhardwaj) [reply csv-extractor/csv_extractor.py]: Everything parsed by csv is stored in a string. Empty/None values are always parsed as empty string. I can confirm and make these changes; however, I think this check is sufficient.

File: csv-extractor/csv_extractor.py (hunk 2 above)
Comment (VibhanshuBhardwaj) [reply csv-extractor/csv_extractor.py]: I used that initially, and it was slower than try/except. Not by much but I didn't want it to be slow :)

File: csv-extractor/csv_extractor.py (hunk 3 above)
Comment (VibhanshuBhardwaj) [reply csv-extractor/csv_extractor.py]: Will fix, thanks

File: csv-extractor/csv_extractor.py (hunk 4 above)
Comment (VibhanshuBhardwaj) [reply csv-extractor/csv_extractor.py]: You're right. This is not ideal.

Another possible solution is we parse a file, and post its datapoints parallely, and then move on to the next one.

Which is slightly slower (0.71 seconds/file vs 0.8 seconds/file)

File: csv-extractor/csv_extractor.py (hunk 1 above)
Comment (eventh) [reply csv-extractor/csv_extractor.py]: With pandas 1,5 would be parsed as a string, while 1 would be parsed to a number. If the csv reader only returns string, that means we can simplify some parts of the following lines.

If everything is a string, and we don't expect "null" to be used, or "NaN", then the best option is just "if not value_string:" imo.

File: csv-extractor/csv_extractor.py (hunk 4 above)
Comment (eventh) [reply csv-extractor/csv_extractor.py]: I think we should go for the safer logic even if it is slightly slower for now.

File: csv-extractor/csv_extractor.py (hunk 1 above)
Comment (VibhanshuBhardwaj) [reply csv-extractor/csv_extractor.py]: I am not sure if I understood you correctly, but I made changes here and looks much better now, thanks 👍

File: csv-extractor/csv_extractor.py (hunk 2 above)
Comment (VibhanshuBhardwaj) [reply csv-extractor/csv_extractor.py]: Done

File: csv-extractor/csv_extractor.py (hunk 3 above)
Comment (VibhanshuBhardwaj) [reply csv-extractor/csv_extractor.py]: Done

File: csv-extractor/csv_extractor.py (hunk 4 above)
Comment (VibhanshuBhardwaj) [reply csv-extractor/csv_extractor.py]: Please take a look at the new logic

**Pull request merged**